### PR TITLE
Example code for pulsar

### PR DESCRIPTION
**Pull request by rshermanTHG**

### CODE REVIEW

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarEndpoint.java [hunk 1]
```diff
@@ -65,4 +65,9 @@ public PulsarClient getPulsarClient() {
     public PulsarEndpointConfiguration getConfiguration() {
         return pulsarEndpointConfiguration;
     }
+
+    @Override
+    protected String createEndpointUri() {
+        return "pulsar:cameltest";
```
Comment (rshermanTHG) [inline components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarEndpoint.java]: need to remove this hardcoded topic

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/configuration/PulsarEndpointConfiguration.java [hunk 2]
```diff
@@ -135,7 +131,10 @@ public void setConsumerNamePrefix(String consumerNamePrefix) {
         this.consumerNamePrefix = consumerNamePrefix;
     }
 
-    public PulsarClient getPulsarClient() {
+    public PulsarClient getPulsarClient() throws PulsarClientException {
+        if (pulsarClient == null) {
+            pulsarClient = new ClientBuilderImpl().serviceUrl("pulsar://localhost:6650").build();
```
Comment (rshermanTHG) [inline components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/configuration/PulsarEndpointConfiguration.java]: need to make this configurable

File: examples/camel-example-route-throttling/pom.xml [hunk 3]
```diff
@@ -32,7 +32,7 @@
         route dynamically based on the flow of messages</description>
         
     <properties>
-    <camel.osgi.export.pkg>org.apache.camel.example.client.*</camel.osgi.export.pkg>
+    <camel.osgi.export.pkg>org.apache.camel.example.pulsar.client.*</camel.osgi.export.pkg>
```
Comment (rshermanTHG) [inline examples/camel-example-route-throttling/pom.xml]: This shouldn't be here bloody auto-refactoring

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java [hunk 4]
```diff
@@ -1,24 +1,69 @@
...
 
 public class PulsarComponent extends DefaultComponent {
 
-    public PulsarComponent(CamelContext context) {
+    private static final Logger LOGGER = LoggerFactory.getLogger(PulsarComponent.class);
+
+    private final PulsarEndpointConfiguration configuration;
+    private final AdminConfiguration adminConfiguration;
+
+    PulsarComponent(CamelContext context, PulsarEndpointConfiguration configuration, AdminConfiguration adminConfiguration) {
```
Comment (AbuahK) [inline components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java]: I assumed this constructor is called internally by the camel context?

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/configuration/AdminConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package org.apache.camel.component.pulsar.configuration;
+
+import org.apache.pulsar.client.impl.conf.ClientConfigurationData;
+
+import java.util.Set;
+
+public class AdminConfiguration extends ClientConfigurationData {
```
Comment (AbuahK) [inline components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/configuration/AdminConfiguration.java:7]: Might be best to extract all admin related configuration such as tenant and namespace into the AdminConfiguration file?

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java [hunk 6]
```diff
@@ -1,24 +1,69 @@
...
     @Override
-    protected Endpoint createEndpoint(String uri, String remaining, Map<String, Object> parameters) throws Exception {
-
-        PulsarEndpointConfiguration configuration = new PulsarEndpointConfiguration(new PulsarUri(uri));
+    protected Endpoint createEndpoint(String uri, String path, Map<String, Object> parameters) throws Exception {
 
         setProperties(configuration, parameters);
 
-        return PulsarEndpoint.create(configuration, configuration.getPulsarClient());
+        ensureNameSpaceAndTenant(path);
```
Comment (AbuahK) [inline components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java]: what happens when the admin config file is null and the tenant and namespace is not supplied in the path?

File: components/camel-pulsar/src/test/java/org/apache/camel/component/pulsar/PulsarComponentTest.java [hunk 7]
```diff
@@ -7,7 +7,7 @@
 
     @Test
```
Comment (AbuahK) [inline components/camel-pulsar/src/test/java/org/apache/camel/component/pulsar/PulsarComponentTest.java:20]: We should properly delete this file

**Review by AbuahK — CHANGES_REQUESTED** [review]
A couple comments on this

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarEndpoint.java (hunk 1 above)
Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarEndpoint.java]: Done

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/configuration/PulsarEndpointConfiguration.java (hunk 2 above)
Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/configuration/PulsarEndpointConfiguration.java]: Done

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java (hunk 4 above)
Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java]: Going to add properties and default constructor to allow both ways of configuring. Spring is moving to constructor only injection and deprecating property autowiring.

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/configuration/AdminConfiguration.java (hunk 5 above)
Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/configuration/AdminConfiguration.java:7]: The ClientConfigurationData is what Pulsar expects to be passed in to configure security etc. for the PulsarAdmin API. I could make ClientConfigurationData a field of AdminConfiguration. What do you think @AbuahK ?

Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/configuration/AdminConfiguration.java:7]: Tenant and namespace are specified by the route and so can't be in the AdminConfiguration.

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java (hunk 6 above)
Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java]: Then a topic is created in the default namespace, in the public tenant. This is the default behaviour and if we wish to create a PR upstream must be preserved.

File: components/camel-pulsar/src/test/java/org/apache/camel/component/pulsar/PulsarComponentTest.java (hunk 7 above)
Comment (rshermanTHG) [reply components/camel-pulsar/src/test/java/org/apache/camel/component/pulsar/PulsarComponentTest.java:20]: Or actually write some tests. Which is what I'll do.

File: components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java (hunk 4 above)
Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org.apache.camel.component.pulsar/PulsarComponent.java]: Done

**Review by AbuahK — APPROVED** [review]

**Pull request merged**